### DISCUSSION TITLE
Building "official" JSON Schema foundational tooling

### DISCUSSION THREAD

**Discussion by jviotti** [json-schema-org/community#113] [Governance and Policy]
At the time of this writing, there are is a significant amount of third-party JSON Schema tooling that deviates from the specification despite the best intentions of their authors given the complexity of JSON Schema. When foundational pieces of JSON Schema code start deviating, higher-level software built on top of this shaky foundation starts deviating as well. This results in a significant amount of confusion around certain features of JSON Schema, adds significant support overhead to the JSON Schema team and even results in entire products that mis-use JSON Schema. 

To break this cycle, I propose that the JSON Schema organization maintains foundational tooling around JSON Schema themselves. This would allow JSON Schema to provide robust foundational blocks that others depend on and steer the world towards a correct use of JSON Schema.

I'd like this thread to serve two purposes: identify the minimum set of foundational blocks that must be built to support higher-level applications and identify the guidelines for a JSON Schema module to attain "official" status.

**Foundational Blocks**

These modules must be low-level and generic utilities that other module or product authors can plug together to make use of JSON Schema on their respective use cases.

Some ideas:

- A JSON Schema validator that produces output according to the specification
- A JSON Schema bundler 
- A JSON Schema "walk" module that visits every subschema (as a foundational block to implement linters, etc)
- A JSON Schema testing framework
- A JSON Schema linter
- A JSON Schema tool that transforms schema definitions across drafts for backwards compatibility reasons (in a lossy manner)
- A JSON Pointer and Relative JSON Pointer implementation

**Attaining Official Status**

I think we should clearly define what are the characteristics that a JSON Schema tool must have to be considered "official". These are some of the questions I can think of:

- Should it support the last draft, or the N last drafts?
- Should it be published under an official `jsonschema` namespace? (in the case of NPM)
- Should it be written in JavaScript? What about other programming languages?
- Should it run on the browser?
- Should it be making use of the official test suite in some way or another?
- Should it have an appointed maintainer?
- What is the expected level of documentation?
- What software license must be adopted?

What do you think?

**Comment by gregsdennis** [comment]
We've had internal discussions around this before, and even quite recently, especially since entertaining the idea of joining a publication group such as IETF or JSF, and decided that we want to focus on writing the specification and maintaining the test suite which ensures conformance, leaving implementations and other tooling to the community.

One of the problems that side out of providing such tooling ourselves is language/framework.  How do we, the maintainers of a specification, decide which language is best to write a reference implementation?  Having such an implementation could sway decisions on features because the chosen language does or does not support a proposal (easily).  We tend toward interoperability, so it's disadvantageous for us to choose one over others.

Many of the other questions you mention were also considerations.

**Reply by smoya** [reply]
I understand building tooling in all of the most popular languages is out of the scope. I can also understand maintaining more than one language can be hard, not only because of time investment but synchronizing features between implementations.

However, choosing **one** language and betting on it for all tools will not only remove that friction but become the foundation of tools that can later be ported by the community. 
For illustration purposes, think about the fact of having an official parser implementation (as an example) written in JS. Nowadays JS is a well-known language, with a really big community, tons of open source projects around, and lots of other examples to follow.
Consider having that tooling as a real usability test of new features. Questions such as "Can this new feature be implemented on tooling?" can be answered before accepting a new feature. 

I might be biased by how we do things at AsyncAPI, but I only see benefits in this approach and what @jviotti is presenting.

**Reply by jviotti** [reply]
As a side node, I think that JavaScript itself can be an input to some Wasm tools in order to get it to work with other languages. I remember watching a talk not long ago (I can't remember the name right now) where a team was using some Wasm tool to create C++ plugins for an API Gateway using JavaScript with a lot of success.

**Reply by Relequestual** [reply]
I hear this is possible with .net too @gregsdennis =D
I still need to investigate this for jsonschema.dev, but I have at least 8 other things to do first...

**Reply by jdesrosiers** [reply]
@smoya Maintaining official implementations makes sense for a company like AsyncAPI, but JSON Schema is not that kind of organization. Obviously, the first hurdle is that JSON Schema doesn't have full time developers to work on this tooling. But, even if we did, I wouldn't want to have "official" implementations.

I don't think it's necessary to have official implementations to get standards compliant implementations. IETF maintains the spec for HTTP, but doesn't produce an implementations. Yet, we don't have a problem with standards compliant web servers. What if IETF had declared Apache the official implementation of HTTP? It would have created a monopoly discouraging the development of other web servers and the innovations they introduced.

Let's look at another example that I think is a little closer to what we have in JSON Schema. In the early days of web browsers, different browsers had a lot of incompatibilities. Those incompatibilities were good for the companies making the browser because if they could win the developers, the developers would develop for one browser and users would need to use that browser to view the site properly. But, what happened was that developers rejected browsers that didn't follow the standards and those non-conforming browsers (IE) died out and were replaced by standards-conforming browsers (Edge) resulting in web sites that were compatible (or relatively easily made compatible) with any browser.

I think JSON Schema has gone through a similar progression. In the early days, there were implementations that were slow, had bugs, or didn't conform to the spec in some way. In the JavaScript space, ajv came along and provided a fast, standards-compliant implementation. Today, most people use ajv. We didn't need to vet it or call it official. The community just chose to use the most standards-compliant implementation without our intervention. Just like with browsers, most actively developed implementations are standards-compliant. Our official test suite is very comprehensive and all the implementations test against it.

Ajv was like the Google Chrome of JSON Schema implementations. Unfortunately, just like Chrome, I feel like ajv today is an effective monopoly that is stifling innovation and not clearly better than alternatives. It has it's own custom features, unimplemented features, and even bugs. But, ajv is so dominant, people would rather use ajv than something more standards-compliant or innovative. And if developers are using some of those custom features, they can't easily migrate to something else.

So, history (including JSON Schema history) tells us that having one implementation elevated above the rest ends up being a negative force for interoperability and innovation. That's why I want JSON Schema to focus on the spec rather than official implementations. We'll help in any way we can, but we don't want to chose favorites.

**Reply by jviotti** [reply]
@jdesrosiers I think you have an interesting point, but I wonder if the dynamics behind those examples have some characteristics that JSON Schema is lacking.

> IETF maintains the spec for HTTP, but doesn't produce an implementations. Yet, we don't have a problem with standards compliant web servers.

Network protocols, including HTTP, exist for the sole reason of inter-operability. You have to, almost by definition, write implementations that require to inter-operate with other implementations of the same standard. Because inter-operability is the essence of the thing, there is a very strong natural forcing function to adhere to the specification. 

In the case of JSON Schema, you rarely need to have i.e. two validators producing the exact same result to achieve your goal. You can stick to your broken world and go a very long way without caring much about inter-operability. The natural inter-operability forcing function of something like a networking protocol is not there for JSON Schema IMHO.

> In the early days of web browsers, different browsers had a lot of incompatibilities. Those incompatibilities were good for the companies making the browser because if they could win the developers, the developers would develop for one browser and users would need to use that browser to view the site properly. But, what happened was that developers rejected browsers that didn't follow the standards and those non-conforming browsers (IE) died out and were replaced by standards-conforming browsers (Edge) resulting in web sites that were compatible (or relatively easily made compatible) with any browser.

This case also seems to have a similar natural inter-operability forcing function. If you write a web-app, your user base expects it to look and feel the same way across a variety of web browsers. This user expectation falls in the shoulders of the developer. If the web-app behaves differently in browser X, users will complain to the developer, and either the developer, the user, or both will get annoyed with browser X. Therefore, browsers receive strong signals to not deviate from the pack. 

Again, in the case of JSON Schema, you don't usually write a schema expecting to run against i.e. multiple validators. As long as it works with the one you use, its fine. You are happy, your users are happy, and the forcing function quickly evaporates.

***

The forcing function is there in some cases where you have to plug together multiple JSON Schema tooling that performs different tasks on top of the same schema, like validation + form generation. But I think the signal is still weaker than in your examples, which I think contributes to this problem.

**Reply by gregsdennis** [reply]
> In the case of JSON Schema, you rarely need to have i.e. two validators producing the exact same result to achieve your goal.

Interoperability has always been a goal of JSON Schema.

To say that it's not a requirement is incorrect.  There are many cases where schemas are used by different implementations over the wire or serialized out of one system then deserialized into another.  Without interoperability, this is impossible.

**Reply by gregsdennis** [reply]
We actually get write a few questions around why something works with one online validator but not another.  Interoperability is the solution to this.

**Reply by jviotti** [reply]
@gregsdennis I agree that interoperability is a goal and a crucial part of JSON Schema. 

What I'm trying to say is that in comparison to other standards like HTTP or browser standards, I feel that JSON Schema has a weaker natural forcing function to achieve interoperability from the point of view of implementers. 

I agree that there is a forcing function in some way or another (like the use case you described or the questions JSON Schema gets), but it does not seem enough to stabilize the ecosystem, at least not at this point given what we observe.

**Reply by jviotti** [reply]
Instead of maintaining a set of tooling, another way out of the problem is to create a stronger interoperability forcing function. @Relequestual your idea of "badges" for the various modules that are promoted on the website (and being more picky there) is an effort in this direction. I wonder if more can be done in this direction. To fix the problem, it seems like we need to create an scenario where an implementation (validators, form generation or any other case) is very annoying to use for the most common case unless it conforms 100% to the spec.

**Reply by gregsdennis** [reply]
> from the point of view of implementers

Careful... A lot of us _are_ implementers (myself included).

> at least not at this point given what we observe

Can you link to studies of those observations?

**Reply by jviotti** [reply]
> Careful... A lot of us are implementers (myself included).

I know that. And the tooling produced by people closely involved in JSON Schema is robust. 

However, tooling produced by the rest of people seems to not be as robust. The difference there is how much do people know about the details of JSON Schema to produce good tooling. People that are not super involved in this community lack the deep understanding needed to produce good tooling. Because things are usually fine even if they don't get such understanding (because even though interoperability is important, they can typically go a long way without it in the context of JSON Schema), then they are also not motivated to learn the details, and a cycle is created. 

> Can you link to studies of those observations?

This post started as an idea from a conversation with @Relequestual , who mentioned `json-schema-ref-parser` in particular and AJV having certain deviating features. I think Ben might have more examples to add to this discussion.

**Reply by jdesrosiers** [reply]
@jviotti I think your premise is wrong.

> in the case of JSON Schema, you don't usually write a schema expecting to run against i.e. multiple validators.

I think this is exactly backwards. While there exist uses cases for JSON Schema that are self-contained to a single application, this is the exception, not the rule. The primary use-case that JSON Schema is designed for is message contracts in APIs. Clients that you don't control will use the schemas to verify messages before sending them and the schemas will also be used on the server to verify messages it receives. Nearly the whole value proposition for JSON Schema is that it's interoperable. If you don't need interoperability, you're almost certainly better off going with something else that isn't limited by the tradeoffs of something designed for interoperability ([ajv vs joi](https://stackoverflow.com/questions/54228539/why-is-joi-more-popular-than-ajv/54230035#54230035)).

> I agree that there is a forcing function in some way or another (like the use case you described or the questions JSON Schema gets), but it does not seem enough to stabilize the ecosystem, at least not at this point given what we observe.

Again, I think your premise is wrong. Much like the history of browsers, in the early days we saw a lot of divergent implementations, but today we have a very stable ecosystem. We have a very comprehensive test suite that all the actively maintained implementations use to ensure they are compliant.

> `json-schema-ref-parser` in particular and AJV having certain deviating features.

json-schema-ref-parser is not a JSON Schema implementation. It's just poorly named. Like Chrome, ajv gets away with a lot due to it's market dominance, but even with ajv, the custom features are opt-in and rarely used. The problems we have with ajv today have more to do with what they don't implement than any customizations. The only implementation I'm aware of that diverges from the spec in a significant way is [Opis JSON Schema](https://opis.io/json-schema/), which we declined to add to our implementations page and has shown no indication of having negative effects in the community. The only implementation out there that has a negative effect on our community is https://jsonschema.net (not to be confused with JsonSchema.Net, @gregsdennis's .Net implementation) that generates invalid schemas leading to misunderstandings.

So, while there are a couple instances of some problematic implementations, they are generally minor and exceptional. It's not a rampant problem that we need "official" implementations to solve.

**Reply by Relequestual** [reply]
tl;dr: I don't want this Discussion to be about having an "official validator" or not, because I'm confident we don't need or want one. I want us to think about OTHER tooling we could build.

> Nearly the whole value proposition for JSON Schema is that it's interoperable. If you don't need interoperability, you're almost certainly better off going with something else that isn't limited by the tradeoffs of something designed for interoperability (ajv vs joi). - @jdesrosiers 

100% agree on this! I raise this often.
"If you need to do more than JSON Schema can do, and interoperability isn't a concern, then you might have the wrong tool for the job."

---

> This post started as an idea from a conversation with @Relequestual , who mentioned json-schema-ref-parser in particular and AJV having certain deviating features. I think Ben might have more examples to add to this discussion. - @jviotti 

I don't think we as an org need to create and publish a JSON Schema Validator. We invest in the test suite. We try to make picking an implementation. We can do more to help this with a badging system. I'm working on a proposal for this.

What I DO think we can consider is other tooling, such as a linter.

### JSON Schema ref parser that isn't?
`json-schema-ref-parser` is IMHO badly named. Personally I want to superseed it with a simple bundler (such as the one @jdesrosiers recently published), but the problem is tooling which doesn't want to handle ANY references is still going to be around.

The problem is, people will continue to write tooling which "works for them" and simple schemas, but that doesn't support references in any way. We should avoid using those or help them support internal references. I think most people underestimate how a basic and simple implementation of using references can work. (This would be an example of a thing we could build as a reference implementation to demonstrate.)

I know @philsturgeon would like us to rally and try to upgrade `json-schema-ref-parser` to be JSON Schema aware, but I don't know if that effort is worthwhile when a new bundler was written overnight that follows the (now defined) bundling process. Granted, it wasn't defined previously. I don't know how much effort it would take, nor if it would result in breaking expectations for current users. (The new bundler looks to do a single thing, while `json-schema-ref-parser` does several.)

(The work @philsturgeon does for open source is admirable and I have a lot of time for him. I do however struggle to get excited about a tool which historically seems to cause lots of problems. Especially when there's a new simpler tool which will be easier to maintain.)

---

> The problems we have with ajv today have more to do with what they don't implement than any customizations.

There is the issue of the current default settings which throw errors when processing schemas AJV deems to be "not best practice", yet are totally valid under the spec. Sure, you can change these options, but only when you're using it directly. If it's a deeply nested dependency, [you're out of luck](https://stackoverflow.com/questions/68419621/a-strict-mode-warning-from-ajv-when-using-electron-store), and that just sucks.

**Reply by jdesrosiers** [reply]
Oh yeah, I can't believe I forgot about ajv's "linting" features. That is divergent and has certainly been causing problems.

As for json-schema-ref-parser, I don't think that's a matter of us rallying. My understanding was that the maintainer is simply not interested in including JSON Schema awareness in the library. The problem isn't that no one has made the effort to do it, it's that the maintainer doesn't want it. I could be wrong about that, but that was my understanding.

**Reply by Relequestual** [reply]
The maintainer has stepped back a little and it's now housed under https://github.com/APIDevTools which is mostly driven by Phil, as I understand.

In recent private DMs, it sounded like Phil wanted to see a scope change to make it JSON Schema aware. Hopefully @philsturgeon will pick up on this discussion and clarify =]

**Reply by philsturgeon** [reply]
@jdesrosiers that's definitely not the case. The threads you've seen were talking about the original intent of the code, not about what is wanted or not. Instead of trying to wrangle json-schema-ref-parser into something that properly supported $id/$schema, the maintainer started creating https://github.com/APIDevTools/json-schema-reader instead. This would have everything folks liked about json-schema-ref-parser, but be properly JSON Schema modern compatible. 

Mid way through this new build, the maintainers family went through some difficulties which I wont get into, but they were forced to refocus their time and abandon the work on both tools. I got @stoplightio to pick up some of the maintenance work on json-schema-ref-parser, but that did not include time to finish the rewrite json-schema-reader.

So, that library is still the north star for solving the problem, and it would definitely benefit a thorough investigation to see if it's any good.

If it's good but needs work, let's find people to get the work done. 

If it's no good, put together a library which does what is aims to do and write a migration guide. I'll help folks transition. 

As I said to Ben before we don't need to fight it or explain to people why its crap, we need to replace it with something viable and help people move over, and maybe that's json-schema-reader and maybe its something you folks knock up.

**Reply by jdesrosiers** [reply]
@philsturgeon Thanks for the correction and additional context.

**Comment by Relequestual** [comment]
I think some of these items would be in scope, and some would be out of scope.
My feeling considerations will be different if we had full time staff to develop and maintain tools. It's highly likely at least one full time developer will be joining next year, and I'm hoping for more.

Let's approach this discussion with an air of "what if?" or "what would it take?" thinking.
Let's imagine.

**Reply by jviotti** [reply]
Indeed. I realize I should have mentioned that. My proposal is based on the assumption that JSON Schema will have N number of full-time developers soon.

**Comment by Relequestual** [comment]
FYI, here's a strawman badges proposal: https://github.com/json-schema-org/community/discussions/118
